core/metrics-core-service/src/main/java/org/hawkular/metrics/core/service/TagQueryParser.java

```diff
@@ -0,0 +1,243 @@
+            costSortedMap.put(GROUP_C_COST, new ArrayList<>());
+
+            for (Map.Entry<String, String> tagQuery : tagsQuery.entrySet()) {
+                if(tagQuery.getKey().startsWith("!")) {
+                    // In-memory sorted query, requires fetching all the definitions
+                    List<Map.Entry<String, String>> entries = costSortedMap.get(GROUP_C_COST);
+                    entries.add(tagQuery);
+                    costSortedMap.put(GROUP_C_COST, entries);
+                } else {
+                    // TODO How to filter exact matching from regexp matching?
+                    List<Map.Entry<String, String>> entries = costSortedMap.get(GROUP_B_COST);
+                    entries.add(tagQuery);
+                    costSortedMap.put(GROUP_B_COST, entries);
+                }
+            }
+
+            return costSortedMap;
+        }
+    }
+
+    public <T> Observable<Metric<?>> findMetricsWithFilters(String tenantId, MetricType<T> metricType,
+                                                            Map<String, String> tagsQueries) {
+
+        Map<Long, List<Map.Entry<String, String>>> costSortedMap = QueryOptimizer.reOrderTagsQuery(tagsQueries);
+
+        List<Map.Entry<String, String>> groupBEntries = costSortedMap.get(QueryOptimizer.GROUP_B_COST);
+        List<Map.Entry<String, String>> groupCEntries = costSortedMap.get(QueryOptimizer.GROUP_C_COST);
+
+        Observable<Metric<?>> groupMetrics;
+
+        // Fetch everything from the tagsQueries
+        groupMetrics = Observable.from(groupBEntries)
+                        .flatMap(e -> dataAccess.findMetricsByTagName(tenantId, e.getKey())
+                                .filter(tagValueFilter(e.getValue(), 3))
+                                .compose(new TagsIndexRowTransformer<>(metricType))
+                                .compose(new ItemsToSetTransformer<>())
+                                .reduce((s1, s2) -> {
+                                    s1.addAll(s2);
+                                    return s1;
+                                }))
+                        .reduce((s1, s2) -> {
+                            s1.retainAll(s2);
+                            return s1;
+                        })
+                        .flatMap(Observable::from)
+                        .flatMap(metricsService::findMetric);
+
+        // I might not have any metrics yet.. if this is the only query
+        if(groupBEntries.isEmpty() && !groupCEntries.isEmpty()) {
+            // Fetch all the available metrics for this tenant
+
+            groupMetrics = dataAccess.findAllMetricsFromTagsIndex()
+                    .compose(new TagsIndexRowTransformer<>(metricType))
+                    .filter(mId -> mId.getTenantId().equals(tenantId))
+                    .flatMap(metricsService::findMetric);
+
+            Observable<Metric<?>> dataMetrics = metricsService.findAllMetrics()
+                    .filter(metricTypeFilter(metricType));
```

Comment:
It does not actually do the same thing, the latter function has a bad name. "findAllMetrics" fetches everything from the data & data_compressed tables (well, distinct partition keys), but it does not fetch everything from metrics_tags_idx (metricIds, which might not have any data - so no partition key).

So you need to combine the results.